File: src/pipeline/checkpoint.ts
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import { PipelineStage, PipelineState, StageOutput } from './types.js';
import { haanSubDir } from '../utils/path-utils.js';
import { logger } from '../utils/logger.js';

export interface CheckpointData {
  task: string;
  mode: 'auto' | 'human';
  completedStages: string[];
  outputs: Record<string, CheckpointStageOutput>;
  retryCount: number;
  improvementPass: number;
  maxRetries: number;
  maxImprovementPasses: number;
  timestamp: number;
}

interface CheckpointStageOutput {
  stage: string;
  success: boolean;
  content: string;
  structured?: unknown;
  duration: number;
  tokensUsed: { input: number; output: number };
}

/**
 * Generate a deterministic hash for a task string (used as checkpoint filename).
 */
function taskHash(task: string): string {
  return crypto.createHash('sha256').update(task).digest('hex').slice(0, 16);
}

/**
 * Get the checkpoints directory, creating it if needed.
 */
function getCheckpointDir(): string {
  const dir = haanSubDir('checkpoints');
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true });
  }
  return dir;
}

/**
 * Save a pipeline checkpoint to disk.
 */
export function saveCheckpoint(state: PipelineState): string {
  const dir = getCheckpointDir();
  const filename = `${taskHash(state.task)}.json`;
  const filepath = path.join(dir, filename);

  const completedStages: string[] = [];
  const outputs: Record<string, CheckpointStageOutput> = {};

  for (const [stage, output] of state.outputs) {
    completedStages.push(stage);
    outputs[stage] = {
      stage: output.stage,
      success: output.success,
      content: output.content,
      structured: output.structured,
      duration: output.duration,
      tokensUsed: output.tokensUsed,
    };
  }

  const data: CheckpointData = {
    task: state.task,
    mode: state.mode,
    completedStages,
    outputs,
    retryCount: state.retryCount,
    improvementPass: state.improvementPass,
    maxRetries: state.maxRetries,
    maxImprovementPasses: state.maxImprovementPasses,
    timestamp: Date.now(),
  };

  fs.writeFileSync(filepath, JSON.stringify(data, null, 2), 'utf-8');
  logger.info('checkpoint', `Saved checkpoint: ${filepath}`);
  return filepath;
}

/**
 * Load a checkpoint from disk.
 * Returns null if not found or invalid.
 */
export function loadCheckpoint(taskOrPath: string): CheckpointData | null {
  let filepath: string;

  if (fs.existsSync(taskOrPath)) {
    filepath = taskOrPath;
  } else {
    const dir = getCheckpointDir();
    filepath = path.join(dir, `${taskHash(taskOrPath)}.json`);
  }

  if (!fs.existsSync(filepath)) {
    return null;
  }

  try {
    const raw = fs.readFileSync(filepath, 'utf-8');
    return JSON.parse(raw) as CheckpointData;
  } catch (err) {
    logger.warn('checkpoint', `Failed to load checkpoint: ${filepath}`, err);
    return null;
  }
}

/**
 * Restore a PipelineState from checkpoint data.
 */
export function restoreFromCheckpoint(data: CheckpointData): {
  state: Partial<PipelineState>;
  completedStages: PipelineStage[];
  outputs: Map<PipelineStage, StageOutput>;
} {
  const outputs = new Map<PipelineStage, StageOutput>();
  const completedStages: PipelineStage[] = [];

  for (const [stageStr, output] of Object.entries(data.outputs)) {
    const stage = stageStr as PipelineStage;
    completedStages.push(stage);
    outputs.set(stage, {
      stage,
      success: output.success,
      content: output.content,
      structured: output.structured,
      duration: output.duration,
      tokensUsed: output.tokensUsed,
    });
  }

  return {
    state: {
      task: data.task,
      mode: data.mode,
      retryCount: data.retryCount,
      improvementPass: data.improvementPass,
      maxRetries: data.maxRetries,
      maxImprovementPasses: data.maxImprovementPasses,
    },
    completedStages,
    outputs,
  };
}

/**
 * Remove a checkpoint after successful completion.
 */
export function removeCheckpoint(task: string): void {
  const dir = getCheckpointDir();
  const filepath = path.join(dir, `${taskHash(task)}.json`);
  try {
    if (fs.existsSync(filepath)) {
      fs.unlinkSync(filepath);
      logger.debug('checkpoint', `Removed checkpoint: ${filepath}`);
    }
  } catch {
    // ignore
  }
}

/**
 * List all available checkpoints.
 */
export function listCheckpoints(): CheckpointData[] {
  const dir = getCheckpointDir();
  if (!fs.existsSync(dir)) return [];

  const files = fs.readdirSync(dir).filter(f => f.endsWith('.json'));
  const checkpoints: CheckpointData[] = [];

  for (const file of files) {
    try {
      const raw = fs.readFileSync(path.join(dir, file), 'utf-8');
      checkpoints.push(JSON.parse(raw) as CheckpointData);
    } catch {
      // skip invalid
    }
  }

  return checkpoints.sort((a, b) => b.timestamp - a.timestamp);
}

File: backend/src/haan/pipeline/checkpoint.py
"""
Pipeline checkpoint save/restore for crash recovery.

Checkpoints are saved as JSON files in the .haan/checkpoints/ directory
after each stage completes. If the pipeline crashes, it can resume from
the last checkpoint instead of re-running completed stages.
"""

from __future__ import annotations

import hashlib
import json
import re
from pathlib import Path
from typing import Any

from haan.pipeline.types import PipelineStage, PipelineState, StageOutput
from haan.utils.logger import logger
from haan.utils.path_utils import haan_sub_dir


def _checkpoint_dir() -> Path:
    """Return the checkpoint storage directory."""
    return Path(haan_sub_dir("checkpoints"))


def _checkpoint_path(task: str) -> Path:
    """
    Generate a deterministic checkpoint file path for a given task.

    Uses a hash of the task description to create a safe filename.
    """
    # Create a safe filename from the task description
    task_hash = hashlib.sha256(task.encode()).hexdigest()[:12]
    safe_name = re.sub(r"[^a-zA-Z0-9]", "_", task[:50])
    return _checkpoint_dir() / f"{safe_name}_{task_hash}.json"


def save_checkpoint(state: PipelineState) -> str:
    """
    Save the current pipeline state as a checkpoint.

    Args:
        state: The current pipeline state to persist.

    Returns:
        Absolute path to the saved checkpoint file.
    """
    path = _checkpoint_path(state.task)

    # Build serializable checkpoint data
    data = {
        "task": state.task,
        "mode": state.mode,
        "retry_count": state.retry_count,
        "improvement_pass": state.improvement_pass,
        "completed_stages": [
            stage.value
            for stage, output in state.outputs.items()
            if output.success
        ],
        "outputs": {
            stage.value: output.to_dict()
            for stage, output in state.outputs.items()
        },
    }

    path.write_text(json.dumps(data, indent=2), encoding="utf-8")
    logger.info("checkpoint", f"Saved checkpoint: {path.name}")
    return str(path)


def load_checkpoint(path: str) -> dict[str, Any] | None:
    """
    Load a checkpoint from a file path.

    Args:
        path: Path to the checkpoint JSON file.

    Returns:
        Checkpoint data dict, or None if the file doesn't exist or is invalid.
    """
    p = Path(path)
    if not p.exists():
        logger.warn("checkpoint", f"Checkpoint not found: {path}")
        return None

    try:
        data = json.loads(p.read_text(encoding="utf-8"))
        logger.info("checkpoint", f"Loaded checkpoint: {p.name}")
        return data
    except (json.JSONDecodeError, IOError) as e:
        logger.error("checkpoint", f"Failed to load checkpoint: {e}")
        return None


def restore_from_checkpoint(
    data: dict[str, Any],
) -> tuple[list[PipelineStage], dict[PipelineStage, StageOutput]]:
    """
    Restore pipeline state from checkpoint data.

    Args:
        data: Checkpoint data as returned by load_checkpoint().

    Returns:
        Tuple of (completed_stages, stage_outputs) for restoring pipeline state.
    """
    completed_stages = [PipelineStage(s) for s in data.get("completed_stages", [])]

    outputs: dict[PipelineStage, StageOutput] = {}
    for stage_str, output_data in data.get("outputs", {}).items():
        stage = PipelineStage(stage_str)
        outputs[stage] = StageOutput.from_dict(output_data)

    return completed_stages, outputs


def remove_checkpoint(task: str) -> None:
    """
    Remove the checkpoint file for a completed task.

    Args:
        task: The task description used to generate the checkpoint filename.
    """
    path = _checkpoint_path(task)
    if path.exists():
        path.unlink()
        logger.info("checkpoint", f"Removed checkpoint: {path.name}")
